Sources/SwiftSyntax/Documentation.docc/Working with SwiftSyntax.md

```diff
@@ -0,0 +1,221 @@
+```
+
+Unexpected nodes allow parsers to gracefully recover during parsing by scanning
+forward in the input stream until they reach a known recovery point. The
+Swift parser included with the swift-syntax package uses this implementation
+strategy to great effect. This makes the parser resilient to a broad 
+class of input errors that users of editors and IDEs would make during the 
+course of an editing session.
+
+## Syntax Nodes
+
+A syntax tree is made up of elements called syntax nodes. To help categorize
+and organize syntax nodes, SwiftSyntax defines a hierarchy of protocols for
+syntax nodes. At the top of this hierarchy is the ``SyntaxProtocol`` protocol.
+To represent related categories of syntax, ``SyntaxProtocol`` is further 
+refined:
+
+- ``DeclSyntaxProtocol`` for declarations like `struct`s, `class`es, `enum`s and `protocol`s.
+- ``StmtSyntaxProtocol`` for statements like `if`, `switch`, and `do`.
+- ``ExprSyntaxProtocol`` for expressions like function calls, literals, and closures
+- ``TypeSyntaxProtocol`` for types like `Array<String>`, `[String: String]`, and `some Protocol`
+- ``PatternSyntaxProtocol`` for patterns like `case (_, let x)`
+
+Syntax nodes form the "branches" of the syntax tree, as they are generally 
+high-level collections of one or more syntax nodes. Taken together, these
+branches form the syntactic structure of the source code. This structure is
+used by compilers and static analyzers to process source code at a high level.
+
+A special kind of syntax node is a ``SyntaxCollection``, which represents syntax
+with a variable number of children. For example, a code block value can contain
+zero or more statements in between a pair of braces. To represent these 
+children, a ``CodeBlockSyntax`` value has a ``CodeBlockSyntax/statements`` 
+accessor that returns a ``CodeBlockItemListSyntax`` value. The elements of this
+syntax collection are ``CodeBlockItemSyntax`` values.
+
+## Syntax Tokens
+
+At the leaves of the syntax tree are ``TokenSyntax`` values. A token syntax 
+value represents an atomic unit of syntax and any associated trivia, like an 
+identifier and the whitespace around it. Taken together, the tokens of a syntax 
+tree represent the lexical structure of the source code. This structure is used
+by linters and formatters to analyze the textual content of source code. 
+
+## Syntax Trivia
+
+Trivia includes whitespace, comments, compiler directives, and erroneous source
+code. Trivia contributes most meaningfully to the lexical structure of a 
+document, but often has little bearing on its syntactic structure. For this
+reason, tools like compilers often discard trivia during source code processing.
+However, maintaining trivia is important for tools like editors, IDEs, 
+formatters, and refactoring engines. SwiftSyntax represents trivia explicitly
+with the ``Trivia`` type. Trivia associated with token syntax can be inspected 
+with  the ``TokenSyntax/leadingTrivia-7p6tp`` and 
```

Comment:
Out of curiosity: Where does `7p6tp` come from?

Reply:
Docc itself. I'm not sure how their mangling scheme works, but it disambiguates two accessors with the same name with different types.

Reply:
Interesting.